mibidata/mibi_image.py

```diff
@@ -95,53 +114,79 @@ class MibiImage():
-        filename: The name of the Run XML file which corresponds to the run
-            name.
+        filename: The name of the file with the run metadata.
+        description: String describing the image.
+        optional_metadata: A dictionary for storing additional metadata.
     """
 
-    # pylint: disable=too-many-arguments
-    def __init__(self, data, channels, run=None, date=None,
-                 coordinates=None, size=None, slide=None, point_name=None,
-                 folder=None, dwell=None, scans=None, aperture=None,
-                 instrument=None, tissue=None, panel=None, version=None,
-                 mass_offset=None, mass_gain=None, time_resolution=None,
-                 miscalibrated=None, check_reg=None,
-                 datetime_format=_DATETIME_FORMAT, filename=None):
+    def __init__(self, data, channels, **kwargs):
 
         self._length = len(channels)
         if data.shape[2] != self._length:
             raise ValueError('Channels length does not match data dimensions.')
         self.data = data
         self._set_channels(channels, self._length)
 
+        # check for required metadata in input arguments
+        for attr in _ATTRIBUTES:
+            if attr != 'optional_metadata':
+                try:
+                    kwargs[attr]
+                except KeyError:
+                    kwargs[attr] = None
+        try:
+            kwargs['datetime_format']
+        except KeyError:
+            kwargs['datetime_format'] = _DATETIME_FORMAT
+
+        # fill required metadata
+        date = kwargs.pop('date')
+        datetime_format = kwargs.pop('datetime_format')
         try:
             self.date = datetime.datetime.strptime(date, datetime_format)
         except TypeError:  # Given as datetime obj already, or None.
             self.date = date
+        self.run = kwargs.pop('run')
+        self.coordinates = kwargs.pop('coordinates')
+        self.size = kwargs.pop('size')
+        self.slide = kwargs.pop('slide')
+        self.fov_id = kwargs.pop('fov_id')
+        self.fov_name = kwargs.pop('fov_name')
+        self.folder = kwargs.pop('folder')
+        self.dwell = kwargs.pop('dwell')
+        self.scans = kwargs.pop('scans')
+        self.aperture = kwargs.pop('aperture')
+        self.instrument = kwargs.pop('instrument')
+        self.tissue = kwargs.pop('tissue')
+        self.panel = kwargs.pop('panel')
+        self.version = kwargs.pop('version')
+        self.mass_offset = kwargs.pop('mass_offset')
+        self.mass_gain = kwargs.pop('mass_gain')
+        self.time_resolution = kwargs.pop('time_resolution')
+        self.miscalibrated = kwargs.pop('miscalibrated')
+        self.check_reg = kwargs.pop('check_reg')
+        self.filename = kwargs.pop('filename')
+        self.description = kwargs.pop('description')
+
+        # whatever remains (if anything) is optional metadata
+        try:
+            optional_metadata = kwargs.pop('optional_metadata')
```

Comment:
I don't like the implementation of optional_metadata here because it requires two different ways of accessing MibiImage metadata, e.g.: `im.run` versus `im.optional_metadata['extra_field_name']`. In addition, when saved to a TIFF, the dictionary will just be saved as a single metadata entry where I think it would be nice if each key value pair in `optional_metadata` was saved as `mibi.key_name: value`.

![image](https://user-images.githubusercontent.com/32075205/68720360-cbf5b480-0563-11ea-8487-8da8abd7035d.png)

I wonder if a solution to this could be saving a list of attribute names at the class level where it defaults to a copy of `_ATTRIBUTES` and adding to it any additional kwargs keys (so that the TIFF writer knows all of the metadata fields present). Then, `__setattr__` and `__getattr__`  could be overridden to allow a user to add his/her own metadata which would be saved and accessed in the same way as defined metadata fields.

Reply:
@jaytarolli I tried adding a list of attribute names as an attribute to the class and overwriting the `setattr()` function in order to store user-defined attribute names in the list, but it's not working. The only way I managed to make this work is reaching to `self.__dict__` in order to get the names of the user-defined attributes. This way, it's possible to add them to the metadata and pass them further down in the code to i.e. the tiff creation. Please have a look and let me know if this works for you.